### PR TITLE
feature: add CRI clean up and restoration

### PR DESCRIPTION
**Pull request by sunpa93**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

-when the driver exits, the CRIs should be cleaned up.
-when the driver restarts, the CRIs should be respawned

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/apis/azuredisk/v1alpha1/types.go
Comment (edreed) [inline pkg/apis/azuredisk/v1alpha1/types.go:100]: sp: `occurred`

File: pkg/controller/volumeattachment.go
Comment (edreed) [inline pkg/controller/volumeattachment.go:97]: There is a race condition here when another thread operates on the same volume attachment concurrently. After release the lock on line 94, the other thread will race to acquire the exclusive lock on line 96. The loser will overwrite the update made by the winner. You should use the [double-checked locking pattern](https://en.wikipedia.org/wiki/Double-checked_locking) here, at least. (Look at the C# example for clarity.)

Rather the acquiring the exclusive lock to update the map entry, use [`atomic.AddInt32`](https://pkg.go.dev/sync/atomic#AddInt32) to update the value.

**Review by edreed — CHANGES_REQUESTED** [review]
Why was charts/latest/azuredisk-csi-driver-v1.4.0.tgz changed? Please revert if unintentional.

File: pkg/controller/volumeattachment.go
Comment (sunpa93) [reply pkg/controller/volumeattachment.go]: I have added the double-checked locking pattern. I left the lock for map entry updates as we want to avoid concurrent writes to the map struct.

File: pkg/controller/pv.go
Comment (edreed) [inline pkg/controller/pv.go:88]: If the element exists, we want to use the current value rather than 1 on line 92 below, so this should be:
~~~
if numRetry, ok = r.retryMap[azVolumeName]; !ok {
    r.retryMap[azVolumeName] = 1
    numRetry = 1
}
~~~

File: pkg/controller/pv.go
Comment (edreed) [inline pkg/controller/pv.go:97]: Since you always end up acquiring the exclusive lock (either to update or delete the element), there's really no benefit to using an RW lock here. Using the `atomic` package functions would make the RW lock worthwhile since the exclusive lock would only be needed in the delete case. I won't block this PR for that change since this is needed to unblock other work, but its something we'll want to do soon.

File: pkg/controller/pv.go
Comment (sunpa93) [reply pkg/controller/pv.go:88]: Ah, good catch. I forgot to move the numTry update into the second check.

**Review by edreed — APPROVED** [review]
One small change needed (https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/pull/85#discussion_r658149850) but otherwise LGTM.

File: pkg/controller/pv.go
Comment (sunpa93) [reply pkg/controller/pv.go:97]: Will do :) I will include this change in controller refactoring PR

File: pkg/controller/common.go
Comment (nearora-msft) [inline pkg/controller/common.go:74]: Should we return from here if errors.IsNotFound(err) is true?

File: pkg/controller/common.go
Comment (nearora-msft) [inline pkg/controller/common.go:98]: Same thing about returning from here if errors.IsNotFound(err) is true

File: pkg/controller/common.go
Comment (sunpa93) [reply pkg/controller/common.go:74]: Yes, that should allow us to skip unnecessary operations :)

File: pkg/controller/pv.go
Comment (nearora-msft) [inline pkg/controller/pv.go:98]: If the element doesn't exist, the value of retry would become 1 on line 89 and it would become 2 on line 98- even for the first retry. Is this intended or should we use _else if numRetry <= maxRetry_  instead of _if numRetry <= maxRetry_ ?

Comment (sunpa93) [reply pkg/controller/pv.go:98]: We should be using else if. Will be pushing the change in shortly.

Comment (sunpa93) [reply pkg/controller/pv.go:98]: On second thought, it should remain as is, given the double-checked locking in place. Instead numRetry should be set to 0 instead of 1 in line 89.

File: pkg/controller/volumeattachment.go
Comment (nearora-msft) [inline pkg/controller/volumeattachment.go:166]: Can't we get the volumeAttachments where volumeAttachment.Spec.Attacher == azureutils.DriverName in the first place instead of filtering them below?

Comment (sunpa93) [reply pkg/controller/volumeattachment.go:166]: In order to do so, `spec.Attacher` needs to be a supported field selector (https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/#supported-fields), which I don't think is.

File: charts/v2.0.0-alpha.1/azuredisk-csi-driver/templates/csi-azuredisk-controller-cleanup.yaml
Comment (nearora-msft) [inline charts/v2.0.0-alpha.1/azuredisk-csi-driver/templates/csi-azuredisk-controller-cleanup.yaml:1]: Can we resuse this template for deleting resources viahelm uninstall for the sanity and integration tests?

File: test/sanity/run-test-v2.sh
Comment (nearora-msft) [inline test/sanity/run-test-v2.sh:26]: Can we reuse the cleanup template for deleting the crds and cris here for test as well? If not, we should make these _kubectl delete_ changes for _test/integration/run-test-v2.sh_ as well.

File: charts/v2.0.0-alpha.1/azuredisk-csi-driver/templates/csi-azuredisk-controller-cleanup.yaml
Comment (sunpa93) [reply charts/v2.0.0-alpha.1/azuredisk-csi-driver/templates/csi-azuredisk-controller-cleanup.yaml:1]: I was hoping to do so but because of different api-server for the created minikube / kind cluster, it is not so straightforward. I am looking for ways to uniformize clean-up methods but it will have to stay this way for now unfortunately.

**Review by nearora-msft — APPROVED** [review]

File: pkg/provisioner/crdprovisioner.go
Comment (nearora-msft) [inline pkg/provisioner/crdprovisioner.go:184]: Shouldn't we make a  azV.UpdateStatus call here after updating the status of the instance?

File: pkg/provisioner/crdprovisioner.go
Comment (nearora-msft) [inline pkg/provisioner/crdprovisioner.go:354]: Same as above. azVA.UpdateStatus() after this?

File: pkg/provisioner/crdprovisioner.go
Comment (sunpa93) [reply pkg/provisioner/crdprovisioner.go:184]: You are right but a regular azV.Update call should do because status is no longer a subresource.

So, we could pull the update call in line 194 out of the if clause.

File: pkg/provisioner/crdprovisioner.go
Comment (sunpa93) [reply pkg/provisioner/crdprovisioner.go:354]: In this case, we wouldn't need a separate update call as we already have a azVA.Update call in line 360.

Comment (sunpa93) [reply pkg/provisioner/crdprovisioner.go:354]: The reason we had a separate updateStatus call previously was that we defined status as a subresource but now that it no longer is, a regular update call should update the status as well.

Comment (edreed) [reply pkg/provisioner/crdprovisioner.go:355]: Why is `Status` no longer a sub-resource? What is the impact of making that change?

Comment (sunpa93) [reply pkg/provisioner/crdprovisioner.go:355]: `Status` is no longer a subresource because with the modification we made to the status struct, it no longer made sense for it to be a pointer variable. And leaving status as a subresource was causing issues with the creation of `AzVolume` and `AzVolumeAttachment` object.

The impact of this change would essentially boil down to what change each `update` call would be making to our CRIs. Previously, it only updated the spec and metadata fields but now it would be updating the status field as well. So we should take this into account when updating objects.

**Pull request merged**
